Optimized Lagrange Interpolation

sidhantnagpal opened the pull request:
#### Brief description of what is fixed or changed
1. `specialpolys.py: interpolating_poly`: The current implementation (`master`) of `interpolating_poly` uses multiplication for numerator `(n - 1)` times per iteration. In this PR, the multiplication for numerator (`n` times) is precalculated and `1` division takes place per iteration.
2. `polyfuncs.py: interpolate`: The current implementation (`master`) of `interpolate` uses an `O(n**2)` method for interpolation in the special case when `X = list(range(1, n + 1))`. In this PR, this is implemented using an `O(n)` method, utilizing the fact that denominator changes by `1` per step.

Benchmarks (for point 1 mainly):
```
In [1]: %paste
from sympy import *
from sympy.abc import x, y, z

def f(n, x, X='x', Y='y'):
    """Construct Lagrange interpolating polynomial for ``n`` data points. """
    if isinstance(X, str):
        X = symbols("%s:%s" % (X, n))

    if isinstance(Y, str):
        Y = symbols("%s:%s" % (Y, n))

    coeffs = []
    numert = Mul(*[ (x - u) for u in X ])

    for i in range(0, n):
        numer = numert/(x - X[i])
        denom = Mul(*[ (X[i] - X[j]) for j in range(0, n) if i != j ])
        coeffs.append(numer/denom)

    return Add(*[ coeff*y for coeff, y in zip(coeffs, Y) ])

def g(n, x, X='x', Y='y'):
    """Construct Lagrange interpolating polynomial for ``n`` data points. """
    if isinstance(X, str):
        X = symbols("%s:%s" % (X, n))

    if isinstance(Y, str):
        Y = symbols("%s:%s" % (Y, n))

    coeffs = []

    for i in range(0, n):
        numer = []
        denom = []

        for j in range(0, n):
            if i == j:
                continue

            numer.append(x - X[j])
            denom.append(X[i] - X[j])

        numer = Mul(*numer)
        denom = Mul(*denom)

        coeffs.append(numer/denom)

    return Add(*[ coeff*y for coeff, y in zip(coeffs, Y) ])

x0, x1, x2, x3, x4, x5, y0, y1, y2, y3, y4, y5 = symbols('x:6, y:6')

## -- End pasted text --

In [2]: %timeit f(5, x)
1000 loops, best of 3: 1.49 ms per loop

In [3]: %timeit g(5, x)
1000 loops, best of 3: 1.7 ms per loop

In [4]: %timeit f(6, x)
100 loops, best of 3: 1.96 ms per loop

In [5]: %timeit g(6, x)
100 loops, best of 3: 2.37 ms per loop
```
#### Other comments

Commit:
Optimized interpolating_poly

Commit:
Add test for symbolic interpolation of degree 4

Commit:
Add tests for numerical interpolation

Commit:
Quadratic to Linear Time Interpolation for List

Commit:
Add test for numerical interpolation using list

jksuom commented on sympy/polys/polyfuncs.py:
```diff
@@ -239,17 +239,26 @@ def interpolate(data, x):
 
-    poly = interpolating_poly(n, x, X, Y)
+            numert = Mul(*[ (x - i) for i in range(1, n + 1) ])
+            denom = Mul(*[ (1 - i) for i in range(2, n + 1) ])
```
This looks like a factorial. Could that be used?

BTW, PEP 8 is not always strictly followed in SymPy, but this rule is mostly applied:

    Avoid extraneous whitespace in the following situations:

    Immediately inside parentheses, brackets or braces.

    Yes: spam(ham[1], {eggs: 2})
    No:  spam( ham[ 1 ], { eggs: 2 } )

sidhantnagpal replied:
> This looks like a factorial. Could that be used?

Yes, with care for negative sign. Will do.

> BTW, PEP 8 is not always strictly followed in SymPy, but this rule is mostly applied.

I remember seeing this type of formatting used in SymPy. 
Should I remove the spaces before `]` and after `[`?

jksuom replied:
I have also seen that type of formatting but that should probably be avoided in new code. (So please remove the spaces.)

sidhantnagpal replied:
> I remember seeing this type of formatting used in SymPy.

In line `138` of `polyfuncs.py`: [here](https://github.com/sidhantnagpal/sympy/blob/lagrange-interpolate/sympy/polys/polyfuncs.py#L138)

sidhantnagpal replied:
> I have also seen that type of formatting but that should probably be avoided in new code. (So please remove the spaces.)

Oh, okay. So should I remove the spaces throughout the file (for existing lines as well)?

Commit:
Use factorial instead of Mul

Commit:
Fixed code formatting

Removed immediate spaces from list comprehension
Added spaces for binary +, - operations
Made range consistent

Commit:
Fixed code formatting

Removed immediate spaces from list comprehension
Made range consistent

sidhantnagpal reviewed (commented):
Made the required changes.

jksuom commented:
Thanks, this looks good.

sidhantnagpal commented:
Thank you @jksuom.
